src/linux/maps_reader.rs

```diff
@@ -130,6 +205,187 @@ impl MappingInfo {
+            return Err(E::InvalidDynamicSectionSize(dynamic_segment_size));
+        }
+
+        let dynamic_section: Vec<Dyn> = memory_reader
+            .read_pod_vec(
+                dynamic_segment_address,
+                dynamic_segment_size / std::mem::size_of::<Dyn>(),
+            )
+            .map_err(E::ReadDynamicSectionFailed)?;
+
+        if dynamic_section.last() != Some(&Dyn::default()) {
+            return Err(E::DynamicSectionMissingTerminator);
+        }
+
+        let debugger_rendezvous_address = dynamic_section
+            .iter()
+            .find_map(|d| (u64::from(d.d_tag) == DT_DEBUG).then_some(d.d_val))
+            .map(|a| usize::try_from(a).unwrap())
+            .ok_or(E::MissingDebugEntry)?;
+
+        let debugger_rendezvous: r_debug = memory_reader
+            .read_pod(debugger_rendezvous_address)
+            .map_err(E::ReadDebuggerRendezvousAddressFailed)?;
+
+        if debugger_rendezvous.r_version != 1 {
+            return Err(E::UnexpectedDebuggerRendezvousVersion(
+                debugger_rendezvous.r_version,
+            ));
+        }
+
+        Self::read_link_map(&mut memory_reader, debugger_rendezvous.r_map)
+            .map_err(|e| E::IterateLinkMapFailed(Box::new(e)))
+    }
+
+    fn read_link_map(
+        memory_reader: &mut MemReader,
+        link_map_address: usize,
+    ) -> std::result::Result<Vec<MappingInfo>, FromDebuggerRendezvousError> {
+        use FromDebuggerRendezvousError as E;
+
+        let mut link: link_map = memory_reader
+            .read_pod(link_map_address)
+            .map_err(E::ReadLinkMapEntryFailed)?;
+
+        if link.l_prev != 0 {
+            return Err(E::InvalidLinkEntry);
+        }
+
+        let mut result: Vec<MappingInfo> = Vec::new();
+
+        loop {
+            let name = {
+                let mut buf = Vec::new();
+                memory_reader
+                    .read_until(link.l_name, 0, &mut buf)
+                    .map_err(E::ReadModuleNameFailed)?;
+                buf.pop();
+                OsString::from(String::from_utf8(buf).unwrap())
+            };
+
+            let module_elf_header_address = link.l_addr;
+            let module_elf_header: ElfHeader = memory_reader
+                .read_pod(module_elf_header_address)
+                .map_err(E::ReadElfHeaderFailed)?;
+            validate_elf_signature(&module_elf_header)?;
+
+            let module_program_header_virtual_address =
+                usize::try_from(module_elf_header.e_phoff).unwrap();
+            let module_program_header_address =
+                module_elf_header_address + module_program_header_virtual_address;
+            let module_program_header_len = usize::from(module_elf_header.e_phnum);
+            let module_program_headers: Vec<ProgramHeader> = memory_reader
+                .read_pod_vec(module_program_header_address, module_program_header_len)
+                .map_err(E::ReadProgramHeaderTableFailed)?;
```

Comment:
As with in the previous function you can break this out into a smaller function to make the logic simpler, e.g. `let module_program_headers = read_module_program_headers(...)?`.